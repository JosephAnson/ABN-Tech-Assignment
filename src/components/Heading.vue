<script setup lang="ts">
import type { PropType } from "vue"

type HeadingSizes = "h1" | "h2" | "h3" | "h4"

const props = defineProps({
  h1: {
    type: Boolean,
    default: false,
  },
  h2: {
    type: Boolean,
    default: false,
  },
  h4: {
    type: Boolean,
    default: false,
  },
  bold: {
    type: Boolean,
    default: false,
  },
  styled: {
    type: String as PropType<HeadingSizes | null>,
    validator: (value: string) => ["h1", "h2", "h3", "h4"].includes(value),
    default: null,
  },
})

const type = computed(() => {
  if (props.h1) return "h1"
  if (props.h2) return "h2"
  if (props.h4) return "h4"

  return "h3"
})

function getHeaderClass(type: HeadingSizes) {
  switch (type) {
    case "h1":
      return "text-2xl md:text-3xl mb-5"
    case "h2":
      return "text-xl md:text-2xl mb-4"
    case "h3":
      return "text-xl md:text-xl mb-3"
    case "h4":
      return "md:text-l mb-2"
  }
}

const styles = computed(() => {
  if (!props.styled) return getHeaderClass(type.value)
  return getHeaderClass(props.styled)
})
</script>

<template>
  <component
    :is="type"
    class="heading font-secondary font-semibold"
    :class="[styles, { 'font-bold': bold }]"
  >
    <slot></slot>
  </component>
</template>
