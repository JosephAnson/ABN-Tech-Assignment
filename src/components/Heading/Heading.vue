<script setup lang="ts">
import type { PropType } from "vue"

type HeadingSizes = "h1" | "h2" | "h3" | "h4"

const props = defineProps({
  h1: {
    type: Boolean,
    default: false,
  },
  h2: {
    type: Boolean,
    default: false,
  },
  h4: {
    type: Boolean,
    default: false,
  },
  bold: {
    type: Boolean,
    default: false,
  },
  styled: {
    type: String as PropType<HeadingSizes | null>,
    validator: (value: string) => ["h1", "h2", "h3", "h4"].includes(value),
    default: null,
  },
})

const type = $computed(() => {
  if (props.h1) return "h1"
  if (props.h2) return "h2"
  if (props.h4) return "h4"

  return "h3"
})

function getHeaderClass(type: HeadingSizes) {
  switch (type) {
    case "h1":
      return "text-3xl pb-5"
    case "h2":
      return "text-2xl pb-4"
    case "h3":
      return "text-xl pb-3"
    case "h4":
      return "text-l pb-2"
  }
}

const styles = $computed(() => {
  if (!props.styled) return getHeaderClass(type)
  return getHeaderClass(props.styled)
})
</script>

<template>
  <component
    :is="type"
    class="heading"
    :class="[styles, { 'font-bold': bold }]"
  >
    <slot></slot>
  </component>
</template>
