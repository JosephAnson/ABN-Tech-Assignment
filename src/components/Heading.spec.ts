import { cleanup, render } from "@testing-library/vue"
import { afterEach, describe, expect, test } from "vitest"
import Heading from "./Heading.vue"

function createRenderHeading(props = {}, text = "Heading") {
  return render(Heading, {
    props,
    slots: {
      default: text,
    },
  })
}

describe("Heading", () => {
  afterEach(() => cleanup())

  test("should render default heading", () => {
    const { html } = render(Heading)
    expect(html()).toMatchSnapshot()
  })

  const headings = ["h1", "h2", "h3", "h4", "p"]
  for (const headingTag of headings) {
    for (const styled of headings) {
      test(`Should display heading ${headingTag} styled as ${styled}`, () => {
        const { html, getByRole, getByText } = createRenderHeading(
          {
            styled,
            h1: headingTag === "h1",
            h2: headingTag === "h2",
            h4: headingTag === "h4",
            p: headingTag === "p",
          },
          `Heading: ${headingTag}, Styled: ${styled}`
        )

        if (headingTag !== "p") {
          expect(
            getByRole("heading", {
              level: parseInt(headingTag.replace("h", "")),
            })
          ).toBeTruthy()
        }

        expect(
          getByText(`Heading: ${headingTag}, Styled: ${styled}`)
        ).toBeTruthy()

        expect(html()).toMatchSnapshot()
      })
    }
  }

  test("should render bold heading", () => {
    const { html, getByText } = createRenderHeading({
      bold: true,
    })
    expect(getByText("Heading").classList.value).toContain("font-bold")
    expect(html()).toMatchSnapshot()
  })
})
