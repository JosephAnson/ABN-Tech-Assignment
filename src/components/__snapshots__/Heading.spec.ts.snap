// Vitest Snapshot v1

exports[`Heading > Should display heading h1 styled as h1 1`] = `"<h1 class=\\"heading font-secondary font-semibold text-2xl md:text-3xl mb-5\\">Heading: h1, Styled: h1</h1>"`;

exports[`Heading > Should display heading h1 styled as h2 1`] = `"<h1 class=\\"heading font-secondary font-semibold text-xl md:text-2xl mb-4\\">Heading: h1, Styled: h2</h1>"`;

exports[`Heading > Should display heading h1 styled as h3 1`] = `"<h1 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: h1, Styled: h3</h1>"`;

exports[`Heading > Should display heading h1 styled as h4 1`] = `"<h1 class=\\"heading font-secondary font-semibold md:text-l mb-2\\">Heading: h1, Styled: h4</h1>"`;

exports[`Heading > Should display heading h1 styled as p 1`] = `"<h1 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: h1, Styled: p</h1>"`;

exports[`Heading > Should display heading h2 styled as h1 1`] = `"<h2 class=\\"heading font-secondary font-semibold text-2xl md:text-3xl mb-5\\">Heading: h2, Styled: h1</h2>"`;

exports[`Heading > Should display heading h2 styled as h2 1`] = `"<h2 class=\\"heading font-secondary font-semibold text-xl md:text-2xl mb-4\\">Heading: h2, Styled: h2</h2>"`;

exports[`Heading > Should display heading h2 styled as h3 1`] = `"<h2 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: h2, Styled: h3</h2>"`;

exports[`Heading > Should display heading h2 styled as h4 1`] = `"<h2 class=\\"heading font-secondary font-semibold md:text-l mb-2\\">Heading: h2, Styled: h4</h2>"`;

exports[`Heading > Should display heading h2 styled as p 1`] = `"<h2 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: h2, Styled: p</h2>"`;

exports[`Heading > Should display heading h3 styled as h1 1`] = `"<h3 class=\\"heading font-secondary font-semibold text-2xl md:text-3xl mb-5\\">Heading: h3, Styled: h1</h3>"`;

exports[`Heading > Should display heading h3 styled as h2 1`] = `"<h3 class=\\"heading font-secondary font-semibold text-xl md:text-2xl mb-4\\">Heading: h3, Styled: h2</h3>"`;

exports[`Heading > Should display heading h3 styled as h3 1`] = `"<h3 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: h3, Styled: h3</h3>"`;

exports[`Heading > Should display heading h3 styled as h4 1`] = `"<h3 class=\\"heading font-secondary font-semibold md:text-l mb-2\\">Heading: h3, Styled: h4</h3>"`;

exports[`Heading > Should display heading h3 styled as p 1`] = `"<h3 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: h3, Styled: p</h3>"`;

exports[`Heading > Should display heading h4 styled as h1 1`] = `"<h4 class=\\"heading font-secondary font-semibold text-2xl md:text-3xl mb-5\\">Heading: h4, Styled: h1</h4>"`;

exports[`Heading > Should display heading h4 styled as h2 1`] = `"<h4 class=\\"heading font-secondary font-semibold text-xl md:text-2xl mb-4\\">Heading: h4, Styled: h2</h4>"`;

exports[`Heading > Should display heading h4 styled as h3 1`] = `"<h4 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: h4, Styled: h3</h4>"`;

exports[`Heading > Should display heading h4 styled as h4 1`] = `"<h4 class=\\"heading font-secondary font-semibold md:text-l mb-2\\">Heading: h4, Styled: h4</h4>"`;

exports[`Heading > Should display heading h4 styled as p 1`] = `"<h4 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: h4, Styled: p</h4>"`;

exports[`Heading > Should display heading p styled as h1 1`] = `"<p class=\\"heading font-secondary font-semibold text-2xl md:text-3xl mb-5\\">Heading: p, Styled: h1</p>"`;

exports[`Heading > Should display heading p styled as h2 1`] = `"<p class=\\"heading font-secondary font-semibold text-xl md:text-2xl mb-4\\">Heading: p, Styled: h2</p>"`;

exports[`Heading > Should display heading p styled as h3 1`] = `"<p class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: p, Styled: h3</p>"`;

exports[`Heading > Should display heading p styled as h4 1`] = `"<p class=\\"heading font-secondary font-semibold md:text-l mb-2\\">Heading: p, Styled: h4</p>"`;

exports[`Heading > Should display heading p styled as p 1`] = `"<p class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\">Heading: p, Styled: p</p>"`;

exports[`Heading > should render bold heading 1`] = `"<h3 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3 font-bold\\">Heading</h3>"`;

exports[`Heading > should render default heading 1`] = `"<h3 class=\\"heading font-secondary font-semibold text-xl md:text-xl mb-3\\"></h3>"`;
